lib/chef/resource/apt_repository.rb

```diff
@@ -319,38 +319,41 @@ def keyring_path
         def install_key_from_uri(key)
           key_name = key.gsub(/[^0-9A-Za-z\-]/, "_")
           keyfile_path = ::File.join(Chef::Config[:file_cache_path], key_name)
+          real_path = keyring_path
           tmp_dir = TargetIO::Dir.mktmpdir(".gpg")
           at_exit { TargetIO::FileUtils.remove_entry(tmp_dir) }
 
           if new_resource.signed_by
-            keyfile_path = keyring_path
-
             directory "/etc/apt/keyrings" do
               mode "0755"
             end
-          end
 
-          declare_resource(key_type(key), keyfile_path) do
-            source key
-            mode "0644"
-            sensitive new_resource.sensitive
-            action :create
-            verify "gpg --homedir #{tmp_dir} %{path}"
-          end
+            declare_resource(key_type(key), keyfile_path) do
+              source key
+              mode "0644"
+              sensitive new_resource.sensitive
+              action :create
+              verify "gpg --homedir #{tmp_dir} %{path}"
+              notifies :run, "execute[dearmor #{real_path}]", :immediately
+            end
 
-          # If signed by is true, then we don't need to
-          # add to the default keyring. Instead make sure it's dearmored
-          if new_resource.signed_by
-            execute "gpg dearmor key" do
-              input ::File.read(keyfile_path)
-              command [ "gpg", "--batch", "--yes", "--dearmor", "-o", keyfile_path ]
+            execute "dearmor #{real_path}" do
+              command [ "gpg", "--batch", "--yes", "--dearmor", "-o", real_path, keyfile_path ]
               default_env true
               sensitive new_resource.sensitive
-              action :run
+              action :nothing
               only_if { ::File.read(keyfile_path).include?("-----BEGIN PGP PUBLIC KEY BLOCK-----") }
               notifies :run, "execute[apt-cache gencaches]", :immediately
             end
           else
+            declare_resource(key_type(key), keyfile_path) do
```

Comment:
given your current variables, this would break, but given my renaming of them, this will work.